### PR TITLE
feat(agent): implement low-level snapshot API

### LINKED ISSUE #37: Implement low-level snapshot api
# Overview

Implement a low-level snapshot API for the TypeScript SDK that allows developers to explicitly capture and restore agent state at any point in time. This replicates the functionality from [Python SDK PR #30](https://github.com/zastrowm/sdk-python/pull/30) with TypeScript-specific adaptations.

## Background

The snapshot API provides capabilities distinct from session management:
- **Point-in-time capture**: Save complete agent state at any moment
- **Arbitrary restore**: Restore to any previously captured state  
- **Context management**: Enable advanced strategies with state modifications

This enables use cases like checkpointing before risky operations, implementing undo/redo functionality, and creating branching conversation flows.

---

## Implementation Requirements

Based on [Python SDK PR #30](https://github.com/zastrowm/sdk-python/pull/30) and [SNAPSHOT.md design doc](https://github.com/zastrowm/sdk-python/blob/agent-tasks/29/SNAPSHOT.md), adapted for TypeScript.

#### Snapshot Interface
```typescript
export interface Snapshot {
  /** String identifying the snapshot type (e.g., "agent"). Strands-owned. */
  type: string
  /** Version string for forward compatibility. Strands-owned. */
  version: string
  /** ISO 8601 timestamp of when snapshot was taken. Strands-owned. */
  timestamp: string
  /** Agent's evolving state (opaque to callers). Strands-owned. */
  data: Record<string, JSONValue>
  /** Application-owned data. Strands does not read or modify this. */
  appData: Record<string, JSONValue>
}
```

#### SnapshotPreset Type
```typescript
export type SnapshotPreset = 'session'
```

#### Serializable Interface (New Pattern)
Create a new interface for objects that can serialize/deserialize their state:
```typescript
export interface StateSerializable {
  getSerializedState(): JSONValue
  loadSerializedState(state: JSONValue): void
}
```
*Note: Final name to be determined during implementation (consider: `Serializable`, `StateSerializable`, `StatePersistable`)*

#### takeSnapshot()
```typescript
takeSnapshot(options: {
  appData?: Record<string, JSONValue>
  include?: SnapshotPreset | string[]
  exclude?: string[]
}): Snapshot
```

#### loadSnapshot()
```typescript
loadSnapshot(snapshot: Snapshot): void
```

### Available Snapshot Fields

| Field | Description | In "session" preset |
|-------|-------------|---------------------|
| `messages` | Conversation history | ✓ |
| `state` | Custom application state (`agent.state`) | ✓ |
| `conversationManagerState` | Internal state of conversation manager | ✓ |
| `systemPrompt` | Agent's system prompt | ✗ |

**Note**: `interruptState` is omitted (TypeScript SDK has no interrupt mechanism).

### Include/Exclude Parameters

- `include="session"` — Preset: messages, state, conversationManagerState
- `include=["field1", "field2"]` — Explicit list of fields
- `exclude=["field1"]` — Fields to exclude (applied after include)
- **Validation**: Either `include` or `exclude` must be specified

### What Does NOT Get Snapshotted (Agent Definition)

- tools
- model
- conversationManager instance
- hooks
- printer

### Helper Functions

```typescript
export const SNAPSHOT_VERSION = '1.0'
export const ALL_SNAPSHOT_FIELDS: readonly string[]
export const SESSION_PRESET_FIELDS: readonly string[]

export function createTimestamp(): string
export function resolveSnapshotFields(include?, exclude?): Set<string>
```

---

### New Files
- `src/agent/snapshot.ts` — Snapshot interface, StateSerializable interface, constants, helper functions

### Files to Modify
- `src/agent/agent.ts` — Add `takeSnapshot()` and `loadSnapshot()` methods
- `src/types/messages.ts` — Add `toData()` method to Message class, `fromData()` static method
- `src/conversation-manager/sliding-window-conversation-manager.ts` — Implement `StateSerializable` interface
- `src/conversation-manager/null-conversation-manager.ts` — Implement `StateSerializable` interface
- `src/index.ts` — Export new types and interfaces

---

## Testing Requirements

Create unit tests in `src/agent/__tests__/snapshot.test.ts`:

### Test Scenarios
1. **Empty agent snapshot**: Take/load snapshot of agent with no messages
2. **Mid-conversation snapshot**: Snapshot after conversation, restore, verify state
3. **Round-trip verification**: Take snapshot → modify agent → load snapshot → verify restored state
4. **App data preservation**: Verify application appData is preserved through save/load
5. **Include/exclude validation**: Test preset and explicit field selection
6. **System prompt handling**: Verify system_prompt inclusion when explicitly requested
7. **Error cases**: Test validation errors (neither include nor exclude specified, invalid fields)

### Test Guidelines
- Make tests completely independent and readable standalone
- Tests should demonstrate the API's developer experience
- Follow existing patterns in `src/agent/__tests__/`

---

### State Boundary

| Evolving State (snapshotted) | Agent Definition (not snapshotted) |
|------------------------------|-----------------------------------|
| messages | tools |
| state | model |
| conversationManagerState | conversationManager instance |
| systemPrompt (optional) | hooks |

### Message Serialization
- Add `toData(): MessageData` instance method to Message class
- Add `static fromData(data: MessageData): Message` method (may already exist as `fromMessageData`)
- Store messages as `MessageData[]` in snapshot for JSON serialization

### Conversation Manager State
- Implement `StateSerializable` interface on conversation managers
- `SlidingWindowConversationManager` should serialize/restore any evolving state
- `NullConversationManager` can return empty object

### Design Principles (from SNAPSHOT.md)
1. Snapshots capture **state (data)**, not **runtime behavior (code)**
2. Use interfaces for easy JSON serialization
3. No state transformation during save or load
4. `type`, `version`, `timestamp`, `data` fields are Strands-owned (opaque to callers)
5. `appData` field is application-owned (Strands does not touch)

---

## Out of Scope

- ~~FileSystemPersister~~ (skip for TypeScript - users implement own persistence)
- ~~Interrupt state~~ (TypeScript SDK has no interrupt mechanism)
- Snapshot in Agent constructor (`new Agent({ snapshot: ... })`)
- Integration tests
- Plugin extensibility for snapshot contributions

---

## Acceptance Criteria

- [ ] `Snapshot` interface defined with type, version, timestamp, data, appData
- [ ] `StateSerializable` interface defined for generic state serialization
- [ ] `SnapshotPreset` type defined
- [ ] `Agent.takeSnapshot()` method implemented with include/exclude support
- [ ] `Agent.loadSnapshot()` method implemented
- [ ] `Message.toData()` method added
- [ ] Conversation managers implement `StateSerializable` interface
- [ ] Helper functions: `createTimestamp()`, `resolveSnapshotFields()`
- [ ] Constants: `SNAPSHOT_VERSION`, `ALL_SNAPSHOT_FIELDS`, `SESSION_PRESET_FIELDS`
- [ ] Unit tests covering all scenarios
- [ ] Tests are readable and demonstrate the API developer experience
- [ ] Code follows repository patterns (type annotations, TSDoc, logging style)
- [ ] Exports added to `src/index.ts`

---

## References

- [Python SDK PR #30](https://github.com/zastrowm/sdk-python/pull/30) — Implementation reference
- [SNAPSHOT.md Design Doc](https://github.com/zastrowm/sdk-python/blob/agent-tasks/29/SNAPSHOT.md) — Design specification
- [Python SDK Issue #29](https://github.com/zastrowm/sdk-python/issues/29) — Original feature request

### PR DESCRIPTION
**Pull request by zastrowm**
## Summary

Implements internal Snapshot API for point-in-time capture and restoration of agent state, enabling use cases like checkpointing, undo/redo, and branching conversation flows.

### New Files
- `src/agent/snapshot.ts` - Core snapshot types and top-level functions

### Modified Files  
- `src/types/messages.ts` - Added `systemPromptToData` function
- `src/index.ts` - No snapshot exports (internal API for now)

## API (Internal)

The snapshot functions are internal implementation details for now. We anticipate opening these up as public Agent methods in a future release after API review.

```typescript
import { takeSnapshot, loadSnapshot } from './agent/snapshot.js'

// Take a snapshot with preset
const snapshot = takeSnapshot(agent, { preset: 'session' })

// Take a snapshot with specific fields
const snapshot = takeSnapshot(agent, { include: ['messages', 'state'] })

// Use preset and exclude
const snapshot = takeSnapshot(agent, { preset: 'session', exclude: ['systemPrompt'] })

// Restore from snapshot  
loadSnapshot(agent, snapshot)
```

### TakeSnapshotOptions
- `preset` - Starting set of fields ('session')
- `include` - Fields to add to preset
- `exclude` - Fields to remove (no error if not present)
- `appData` - Application-owned data

### Snapshot Fields
- `messages` - Conversation history
- `state` - AgentState key-value storage  
- `conversationManagerState` - Conversation manager internal state
- `systemPrompt` - System prompt (string or complex content block array)

## Testing
- 29 tests with deterministic time mocking (`vi.setSystemTime`)
- Round-trip serialization tests
- Edge cases (empty state, complex content blocks)

Resolves #37

### CODE REVIEW

File: src/agent/snapshot.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  /**
+   * Application-owned data. Strands does not read or modify this.
+   */
+  appData: Record<string, JSONValue>
+}
+
+/**
+ * Interface for objects that can serialize and deserialize their state.
+ */
+export interface StateSerializable {
```
Comment (zastrowm) [inline src/agent/snapshot.ts]: Put this somewhere non-snapshot specific

File: src/agent/__tests__/snapshot.test.ts [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+      agent = createTestAgent()
+    })
+
+    it('creates snapshot with session preset', () => {
+      agent.messages.push(new Message({ role: 'user', content: [new TextBlock('Hello')] }))
+      agent.state.set('key', 'value')
+
+      const snapshot = agent.takeSnapshot({ include: 'session' })
+
+      expect(snapshot.type).toBe('agent')
```
Comment (zastrowm) [inline src/agent/__tests__/snapshot.test.ts]: Because this is all mock data (or can be mocked), expect on the entire object instead of individual fields

File: src/types/media.ts [hunk 3]
```diff
@@ -270,6 +296,29 @@ export class VideoBlock implements VideoBlockData {
     }
     throw new Error('Invalid video source')
   }
+
+  /**
+   * Converts this VideoBlock to a plain data representation.
+   *
+   * @returns VideoBlockData representation
+   */
+  public toData(): VideoBlockData {
```
[the remark this answers is not in the document]
Comment (zastrowm) [reply src/types/media.ts]: Yes, let's go with this simplified approach for now.  Let's remove toData from all the classes if possible and just use `JSON.parse(JSON.stringify())` for json serialization purposes. AFAIK we should then be able to deserialize by using Message.fromData() to rehydrate.  LMK if I'm wrong
